### DISCUSSION TITLE
How to properly use Ref.refresh(provider.future)? It throws when AsyncValue is AsyncError

### DISCUSSION THREAD

**Discussion by JanHitako** [rrousselGit/riverpod#3441] [Q&A]
I want to implement a pull to refresh. On all example I have seen with riverpod, the recommended method is:

```
return RefreshIndicator(
      onRefresh: () => ref.refresh(feedControllerProvider
          .future), 
      child: CustomScrollView(....) ,
    );
```

The issue with this method while used on my provider: 
- in the case the refresh fail (for example, at network interuption happen when the value of the provider is being re-fetched, this will throw on me:

> [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: AppwriteException: , Le délai de temporisation de sémaphore a expiré.
> 
>  (0)
> #0      ClientIO.call (package:appwrite/src/client_io.dart:381:7)
> <asynchronous suspension>
> #1      Databases.listDocuments (package:appwrite/services/databases.dart:30:17)
> <asynchronous suspension>
> #2      FeedRepository.getFeed (package:news_from_sg/data/feed_repository.dart:15:20)
> <asynchronous suspension>
> #3      FeedController.build (package:news_from_sg/data/feed_controller.dart:14:18)
> <asynchronous suspension>
> #4      FutureHandlerProviderElementMixin.handleFuture.<anonymous closure>.<anonymous closure> (package:riverpod/src/async_notifier/base.dart:348:9)
> <asynchronous suspension>

wherea in any other cases (fail on first load, fail when provider is not listened anymore and then getting listened again, etc), this same exact error is handled perfectly by riverpod.

>  [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: AppwriteException: , Le délai de temporisation de sémaphore a expiré.

But based on this:

> #4      FutureHandlerProviderElementMixin.handleFuture.<anonymous closure>.<anonymous closure> (package:riverpod/src/async_notifier/base.dart:348:9)

I think the issue is about the future in:

> ref.refresh(feedControllerProvider .future)

However, I don't know how to catch it, a simple wrapping of this in a try catch block doesn't work, the code never goes to the catch block and the error still throw.

Would love some insight.

**Comment by JanHitako** [comment]
Solved, for anyone having the same situation : my try catch ended before because I obviously forgot to await for what I have been trying to catch.

**Comment by JanHitako** [comment]
Reopening because the implementation I made actually doesn't throw error but doesn't work either.


First, what is said implementation?
I replaced this :
```
 onRefresh: () => ref.refresh(feedControllerProvider.future)
```

By this:
```
onRefresh: () async {
        try {
          return await ref.refresh(feedControllerProvider.future);
        } catch (_) {}
      }
```

The issue now is that the provider doesn't seem to actually invalidate at all. 
I mean, what I expect is something like this:
we call ref.refresh => the state is disposed then we fetch a new state => the refresh indicator shows and wait for the response => the refresh fail (intentionally) => we enter into state.when(error) => ...

what actually happens:
state isn't disposed, I can still see the value of the previous state in the variables section during debugging during the entire process => isLoading does go to true but that's about it => then we still have an AsyncValue and go to state.when(data) at the end with the state that should have been already disposed. 

Would like some help on this one.

**Reply by AhmedLSayed9** [reply]
AsyncLoading state storing the previous value is a feature.
It enables you to show old data while the refresh indicator is loading.

You can disable it if you want: `asyncValue.when( skipLoadingOnRefresh: false )`
